9-type-aliases-redux/latex/kind-judgements.tex

```diff
@@ -0,0 +1,414 @@
+\newcommand{\kequiv}[3]{#1 \vdash #2 \equiv #3}
+\newcommand{\kcequiv}[4]{#1 \vdash #2 \equiv #3 : #4}
+\newcommand{\ksubkind}[3]{#1 \vdash #2 <:: #3}
+\newcommand{\tyvarValid}[1]{#1 \textt{ valid} }
+\newcommand{\tyvarNotValid}[1]{#1 \textt{ }\lnot\textt{valid} }
+\newcommand{\dexpTypeAssign}[3]{#1 \vdash #2 : #3}
+
+
+\newcommand{\hPhi}{\Phi}
+\newcommand{\hGamma}{\Gamma}
+\newcommand{\EmptyhPhi}{\cdot}
+\newcommand{\EmptyDelta}{\cdot}
+
+\newcommand{\hexp}{\hat{e}}
+\newcommand{\dexp}{e}
+\newcommand{\htau}{\hat{\tau}}
+\newcommand{\hkappa}{\kappa}
+\newcommand{\dtau}{\tau}
+\newcommand{\hrho}{\hat{\rho}}
+
+\newcommand{\Ty}{\textt{Ty}}
+\newcommand{\KHole}{\textt{KHole}}
+\newcommand{\KSing}[1]{\textt{S(}#1{\textt{)}}}
+\newcommand{\hlist}[1]{\textt{list(}#1\textt{)}}
+
+\newcommand{\llparenthesiscolor}{\textcolor{violet}{\llparenthesis}}
+\newcommand{\rrparenthesiscolor}{\textcolor{violet}{\rrparenthesis}}
+\newcommand{\hthole}[1]{\llparenthesiscolor\rrparenthesiscolor^{#1}}
+\newcommand{\hhole}[2]{\llparenthesiscolor#1\rrparenthesiscolor^{#2}}
+\newcommand{\kehole}{\llparenthesiscolor\rrparenthesiscolor}
+
+\newcommand{\idof}[1]{\mathsf{id}(#1)}
+\newcommand{\Dbinding}[3]{#1 :: #3[#2]}
+
+\newcommand{\htdefine}[3]{\textt{type }#1 \textt{ = } #2\textt{ in }#3}
+\newcommand{\dtdefine}[2]{\textt{type bindings } #1 \textt{ in } #2}
+
+\begin{document}
+
+\begin{figure}[t]
+	$\arraycolsep=4pt\begin{array}{rllllll}
+			\mathsf{BinOp}               & \oplus & ::= &
+			\textt{Product} ~\vert~ \textt{Sum} ~\vert~ \textt{Arrow}                                          \\
+			\mathsf{Kind}                & \kappa & ::= &
+			\Ty ~\vert~ \KHole ~\vert~ \KSing{\tau}                                                            \\
+			\mathsf{Constant Types}      & c      & ::= &
+			\textt{Int} ~\vert~ \textt{Float} ~\vert~ \textt{Bool}                                             \\
+			\mathsf{User HTyp}           & \htau  & ::= &
+			c ~\vert~ \htau_1 \oplus \htau_2 ~\vert~ \hlist{\htau} ~\vert~ \hthole{u} ~\vert~ \hhole{\htau}{u} \\
+			\mathsf{Internal HTyp}       & \dtau  & ::= &
+			c ~\vert~ \dtau_1 \oplus \dtau_2 ~\vert~ \hlist{\dtau} ~\vert~ \hthole{u} ~\vert~ \hhole{\dtau}{u} \\
+			\mathsf{Type Vars}           & t      &     &                                                      \\
+			\mathsf{User Type Pattern}   & \hrho  & ::= &
+			t ~\vert~ \hthole{u} ~\vert~ \hhole{t}{u}                                                          \\
+			\mathsf{User Expression}     & \hexp  & ::= &
+			\htdefine{\hrho}{\htau}{\hexp} ~\vert~ elided                                                      \\
+			\mathsf{Internal Expression} & \dtau  & ::= &
+			\dtdefine{\hPhi}{\dexp} ~\vert~ elided                                                             \\
+		\end{array}$
+\end{figure}
+
+
+\begin{minipage}{\linewidth}
+	\judgbox
+	{\kconsubkind{\Delta;\hPhi}{\hkappa_1}{\hkappa_2}}
+	{$\hkappa_1$ is a consistent subkind of $\hkappa_2$}
```

Comment:
i think we need a transitivity rule here?

Reply:
Consistent subkinding (and subtyping) isn’t transitive. Counter example:
S(Int) <~ ?
? <~ S(Bool)

but this is not true:
S(Int) <~ S(Bool)

Reply:
Ah yeah, whoops.

Do you need a rule for S(t1) <~ S(t2) where t1 <~ t2?

Reply:
I think not yet (until polymorphism) because currently:
1. S(tau) is only well-formed if tau has kind Ty
2. forall types tau_1 and tau_2, tau_1 <~ tau_2 iff tau_1 === tau_2 -- and this is covered by the `KCRespectEquiv` rule